### PR TITLE
feat: added the device & cloud destinations for the first release.

### PR DESCRIPTION
**Pull request by hanoak20**
Added Contentstack's browser and cloud destinations.

Description:
For our first release, we're going ahead with a basic setup to sync traits, audience, etc. In the browser plugin, we're just syncing custom traits and events. And in cloud destination, we're syncing computed traits and audiences. We'd follow up on this integration to improve it after we're good with the first release. 
Thanks.

### CODE REVIEW

File: packages/browser-destinations/destinations/contentstack-browser-plugins/src/customAttributesSync/index.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  },
+  perform: (_sdk, { payload, settings }) => {
+    const { traits, userId } = payload
+
+    return fetch(`${settings.personalizeEdgeApiBaseUrl}/user-attributes`, {
+      method: 'PATCH',
+      body: JSON.stringify(traits),
+      headers: {
+        'x-cs-eclipse-user-uid': userId ?? '',
+        'x-project-uid': settings.personalizeProjectId
```
Comment (sankartn) [inline packages/browser-destinations/destinations/contentstack-browser-plugins/src/customAttributesSync/index.ts:35]: 'x-project-uid': settings?.personalizeProjectId

File: packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/index.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      method: 'POST',
+      body: JSON.stringify([
+        {
+          eventKey: event,
+          type: 'EVENT'
+        }
+      ]),
+      headers: {
+        'x-cs-eclipse-user-uid': userId ?? '',
+        'x-project-uid': settings.personalizeProjectId
```
Comment (sankartn) [inline packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/index.ts:40]: 'x-project-uid': settings?.personalizeProjectId

File: packages/destination-actions/src/destinations/contentstack/audienceSync/index.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    }
+  },
+  perform: (request, { payload, settings }) => {
+    const traits = (payload.type === 'track' ? payload.properties : payload.traits) ?? {}
+    return request(`${settings.personalizeEdgeApiBaseUrl}/user-attributes`, {
+      method: 'patch',
+      json: traits,
+      headers: {
+        'x-cs-eclipse-user-uid': payload.userId ?? '',
+        'x-project-uid': settings.personalizeProjectId ?? ''
```
Comment (sankartn) [inline packages/destination-actions/src/destinations/contentstack/audienceSync/index.ts:44]: 'x-project-uid': settings?.personalizeProjectId ?? ''

**Comment by sankartn** [comment]
Hi @joe-ayoub-segment Please review the PR and help us with deployment

**Comment by joe-ayoub-segment** [comment]
Thanks @hanoak20 - I'll schedule this for review.

**Comment by joe-ayoub-segment** [comment]
hi @hanoak20 thanks for raising this PR. 

Looks like some large changes. 

I do have some architectural questions to discuss with you - as there's some design decisions in here which I'm unsure about. However I can deploy it and you can see how it looks in production if you like?

Please note that there will now be 2 Destinations. 

1. The server side Destination with 2 Actions
2. The client side Destination with 2 Actions. 

The customer will need to set each of these Destinations up separately. 

Best regards,
Joe

**Review by joe-ayoub-segment — APPROVED** [review]
Please see comment

File: packages/browser-destinations/destinations/contentstack-browser-plugins/src/customAttributesSync/index.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  defaultSubscription: 'type = "track" or type = "identify"',
+  platform: 'web',
+  fields: {
+    traits: {
+      type: 'object',
+      default: { '@path': '$.traits' },
+      label: 'User traits',
+      description: 'User Profile traits to send to Contentstack',
+      required: true
+    },
```
Comment (joe-ayoub-segment) [inline packages/browser-destinations/destinations/contentstack-browser-plugins/src/customAttributesSync/index.ts:18]:
```suggestion
    traits: {
      type: 'object',
       default: {
        '@if': {
          exists: { '@path': '$.traits' },
          then: { '@path': '$.traits' },
          else: { '@path': '$.context.traits' }
        }
      },
      label: 'User traits',
      description: 'User Profile traits to send to Contentstack',
      required: true
    },
```

File: packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/index.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  description: 'Sync Events to your Contentstack Experience.',
+  defaultSubscription: 'type = "track" or type = "identify"',
+  platform: 'web',
+  fields: {
+    userId: {
+      type: 'string',
+      default: { '@path': '$.userId' },
+      label: 'User ID',
+      description: 'ID for the user',
+      required: false
```
Comment (joe-ayoub-segment) [inline packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/index.ts:17]:
```suggestion
      required: true
```

File: packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/index.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    return fetch(`${settings.personalizeEdgeApiBaseUrl}/events`, {
+      method: 'POST',
+      body: JSON.stringify([
+        {
+          eventKey: event,
+          type: 'EVENT'
+        }
+      ]),
+      headers: {
+        'x-cs-eclipse-user-uid': userId ?? '',
```
Comment (joe-ayoub-segment) [inline packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/index.ts:39]:
```suggestion
        'x-cs-eclipse-user-uid': userId,
```

File: packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/index.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
+import type { BrowserActionDefinition } from '@segment/browser-destination-runtime/types'
+import type { Settings } from '../generated-types'
+import type { Payload } from './generated-types'
+
+// Change from unknown to the partner SDK types
+const action: BrowserActionDefinition<Settings, {}, Payload> = {
+  title: 'Events Sync',
+  description: 'Sync Events to your Contentstack Experience.',
+  defaultSubscription: 'type = "track" or type = "identify"',
```
Comment (joe-ayoub-segment) [inline packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/index.ts:9]:
```suggestion
  defaultSubscription: 'type = "track"',
```

File: packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/generated-types.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+// Generated file. DO NOT MODIFY IT BY HAND.
+
+export interface Payload {
+  /**
+   * ID for the user
+   */
+  userId?: string
```
Comment (joe-ayoub-segment) [inline packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/generated-types.ts:7]:
```suggestion
  userId: string
```

File: packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/generated-types.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+  /**
+   * User Event
+   */
+  event?: string
```
Comment (joe-ayoub-segment) [inline packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/generated-types.ts:11]:
```suggestion
  event: string
```

File: packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/index.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+      label: 'User ID',
+      description: 'ID for the user',
+      required: false
+    },
+    event: {
+      type: 'string',
+      default: { '@path': '$.event' },
+      label: 'User Event',
+      description: 'User Event',
+      required: false
```
Comment (joe-ayoub-segment) [inline packages/browser-destinations/destinations/contentstack-browser-plugins/src/eventsSync/index.ts:24]:
```suggestion
      required: true
```

File: packages/browser-destinations/destinations/contentstack-browser-plugins/src/index.ts [hunk 11]
```diff
@@ -1,21 +1,35 @@
...
 import type { BrowserDestinationDefinition } from '@segment/browser-destination-runtime/types'
 import { browserDestination } from '@segment/browser-destination-runtime/shim'
-import contentstackPlugin from './contentstackPlugin'
+import customAttributesSync from './customAttributesSync'
+import eventsSync from './eventsSync'
 
 // Switch from unknown to the partner SDK client types
 export const destination: BrowserDestinationDefinition<Settings, {}> = {
   name: 'Contentstack Browser Plugins',
+  slug: 'contentstack-browser-plugins',
```
Comment (joe-ayoub-segment) [inline packages/browser-destinations/destinations/contentstack-browser-plugins/src/index.ts:10]:
```suggestion
  name: 'Contentstack Web',
  slug: 'contentstack-web',
```

File: packages/browser-destinations/destinations/contentstack-browser-plugins/src/index.ts [hunk 12]
```diff
@@ -1,21 +1,35 @@
...
   mode: 'device',
-
-  settings: {},
-
+  description: 'Sync data from Segment to Contentstack',
```
Comment (joe-ayoub-segment) [inline packages/browser-destinations/destinations/contentstack-browser-plugins/src/index.ts:12]:
```suggestion
  description: 'Sync web event and user profile data from Segment to Contentstack',
```

File: packages/destination-actions/src/destinations/contentstack/audienceSync/index.ts [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+      label: 'Event Type',
+      description: 'Type of the event',
+      required: true
+    },
+    userId: {
+      type: 'string',
+      default: { '@path': '$.userId' },
+      label: 'User ID',
+      description: 'ID for the user',
+      required: false
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/contentstack/audienceSync/index.ts:34]:
```suggestion
      required: true
```

File: packages/destination-actions/src/destinations/contentstack/audienceSync/generated-types.ts [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    [k: string]: unknown
+  }
+  /**
+   * Type of the event
+   */
+  type: string
+  /**
+   * ID for the user
+   */
+  userId?: string
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/contentstack/audienceSync/generated-types.ts:23]:
```suggestion
  userId: string
```

File: packages/destination-actions/src/destinations/contentstack/computedAttributeSync/index.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+      label: 'Event Type',
+      description: 'Type of the event',
+      required: true
+    },
+    userId: {
+      type: 'string',
+      default: { '@path': '$.userId' },
+      label: 'User ID',
+      description: 'ID for the user',
+      required: false
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/contentstack/computedAttributeSync/index.ts:34]:
```suggestion
      required: true
```

File: packages/destination-actions/src/destinations/contentstack/computedAttributeSync/generated-types.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    [k: string]: unknown
+  }
+  /**
+   * Type of the event
+   */
+  type: string
+  /**
+   * ID for the user
+   */
+  userId?: string
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/contentstack/computedAttributeSync/generated-types.ts:23]:
```suggestion
  userId: string
```

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]

**Review by joe-ayoub-segment — APPROVED** [review]

**Review by joe-ayoub-segment — APPROVED** [review]

**Review by joe-ayoub-segment — APPROVED** [review]

**Pull request closed without merging**